fix(NewTab): [MC-76] Content should roll over at 3am local time

mmiermans opened the pull request:
# Goal
New Tab content should roll over at 3am local time, based on the timezone associated with the NewTab recommendation surface.

In the previous implementation, it incorrectly rolled over at midnight local time. This bug was introduced when migrating from the feed-machine to recommendation-api. Before the migration, content rolled over at 3am local time, as expected.

## Reference

I added to the Firefox NewTab slate spec:
> Content rolls over at 3am local time.
* https://getpocket.atlassian.net/wiki/spaces/PE/pages/2927100008/Fx+NewTab+Slate+spec

Tickets:
* https://mozilla-hub.atlassian.net/jira/software/c/projects/MC/boards/720?modal=detail&selectedIssue=MC-76&assignee=64065e8893cf2599462fe21d

Slack thread:
> Carolyn: The stories you schedule for each day in the Curation Tool will start hitting New Tab at 3am in your New Tab’s local timezone. (For U.S. New Tab that means 3am EST).
* https://mozilla.slack.com/archives/C05EVRU1U1X/p1694016926870279

## Implementation Decisions
- In the test, I made the frozen time a parameter in the test, such that it can be precisely set to just before and after the publishing day is supposed to roll over. That's less confusing and more precise.

rolf-moz commented:
I think there should be some clarification in the spec and communication with the editorial team that we are not rolling over at 'local' time but are instead rolling over at 3am of the time of the associated slate.

in corpus_api_client.py we have a TODO
        # TODO: Modify curated-corpus-api to get timezone from query. Timezones are already hardcoded there.

If this todo is done then we can more closely adhere to the spec. If we can't then we should make clear to the product team that we're moving all cycle times to 3AM slate time, which in the case of EN-US is midnight pacific time and 3AM NY time. So I would make that clear to them first before commuting.

One alternative with less code change is to update get_surface_timezone with a custom time zone for New York + 3h, so the change only affects the US and is more compact.

mmiermans commented:
Great call-outs. They made me realize the PR description was missing several facts:

1. I clarified that 'local time' here refers to the time based on the timezone associated with the slate:
2. I forgot to reference [the NewTab slate spec](https://getpocket.atlassian.net/wiki/spaces/PE/pages/2927100008/Fx+NewTab+Slate+spec). I corrected the desired behavior there:
   > Items scheduled for today or yesterday in the Curation Admin Tool are used as candidate items. Content rolls over at 3am local time.

3. I also quoted the desired behavior that Carolyn posted on Slack:

   > The stories you schedule for each day in the Curation Tool will start hitting New Tab at 3am in your New Tab’s local timezone. (For U.S. New Tab that means 3am EST).

4. I added that this was previously working as expected:

   > This bug was introduced when migrating from the feed-machine to recommendation-api. Before the migration, content rolled over at 3am local time, as expected.
   
> in corpus_api_client.py we have a TODO
> "TODO: Modify curated-corpus-api to get timezone from query. Timezones are already hardcoded there."
> If this todo is done then we can more closely adhere to the spec.

Which spec are you referring to here? The only spec I'm aware of is the [the NewTab slate spec](https://getpocket.atlassian.net/wiki/spaces/PE/pages/2927100008/Fx+NewTab+Slate+spec), which has been updated with the expected behavior. It's definitely not great to have this hardcoded data in two places. It makes a lot more sense to me to put the responsibility of knowing when content rolls over in Curated Corpus API, instead of on the caller (recommendation-api).

> If we can't then we should make clear to the product team that we're moving all cycle times to 3AM slate time, which in the case of EN-US is midnight pacific time and 3AM NY time. So I would make that clear to them first before commuting.

Sorry about the confusion. Is this answered, now that I've added a reference to the expected behavior, as requested by Carolyn, and that we're restoring behavior to what it was prior to the migration away from feed-machine in August?

mmiermans commented on app/data_providers/corpus/corpus_api_client.py:
```diff
@@ -35,8 +35,8 @@ async def fetch(
             }
         """
 
-        # The date is supposed to progress at midnight CET (Central European Time).
```
Yes, it was incorrect before.
